tests/expose_test.go

```diff
@@ -407,6 +407,72 @@ var _ = Describe("Expose", func() {
 
 				Eventually(getStatus, 60, 1).Should(Equal(k8sv1.PodSucceeded))
 			})
+
+			It("Should verify the exposed service is functional before and after VM restart.", func() {
+				vm_obj := vm
+
+				By("Getting back the service's allocated cluster IP.")
+				svc, err := virtClient.CoreV1().Services(vm_obj.Namespace).Get(serviceName, k8smetav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				serviceIP := svc.Spec.ClusterIP
+
+				By("Starting a pod which tries to reach the VMI via ClusterIP.")
+				job := tests.NewHelloWorldJob(serviceIP, servicePort)
+				job, err = virtClient.CoreV1().Pods(vm_obj.Namespace).Create(job)
+				Expect(err).ToNot(HaveOccurred())
+
+				By("Waiting for the pod to report a successful connection attempt.")
+				getStatus := func() k8sv1.PodPhase {
+					pod, err := virtClient.CoreV1().Pods(job.Namespace).Get(job.Name, k8smetav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					return pod.Status.Phase
+				}
+
+				Eventually(getStatus, 60, 1).Should(Equal(k8sv1.PodSucceeded))
+
+				By("Restarting the running VM.")
+				virtctl := tests.NewRepeatableVirtctlCommand("restart", "--namespace", vm.Namespace, vm.Name)
+				err = virtctl()
+				Expect(err).ToNot(HaveOccurred())
+
+				By("Verifying the VMI is running.")
+				Eventually(func() bool {
+					vm_obj, err = virtClient.VirtualMachine(vm_obj.Namespace).Get(vm_obj.Name, &k8smetav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					return vm_obj.Status.Ready
+				}, 120*time.Second, 1*time.Second).Should(BeTrue())
+
+				By("Getting the running VMI.")
+				var vmi *v1.VirtualMachineInstance
+				Eventually(func() bool {
+					vmi, err = virtClient.VirtualMachineInstance(vm_obj.Namespace).Get(vm_obj.Name, &k8smetav1.GetOptions{})
```

Comment:
There's a race condition here. 

I think it's possible for you to be retrieving the old VMI here before the restart.

You should retrieve the vmi object from the first run and save the vmi.UUID in a variable. Then in this loop you should wait until you get a vmi object with a different UUID from the first vmi.

When you issue the "restart" command, behind the scenes the VMI is just being deleted. The VM controller then spins up a new VMI to replace the old VMI that was shut down.  All this happens asynchronously, so you'll have to wait for the UUID of the VMI to change before you can be sure the restart has occurred on the backend.

Reply:
Great point, @davidvossel !
I suspect this race condition failed my CI on the previous commits.
I now handle this race, but differently than what you suggested:
 - Stop the VMI.
 - Verify the VMI is stopped.
 - Start the VMI back.
 - Retrieve the new VMI, hence the new UUID.
Thanks!